src/foam/core/Validation.js

```diff
@@ -15,15 +15,156 @@
  * limitations under the License.
  */
 
+foam.CLASS({
+  package: 'foam.core',
+  name: 'ValidationPredicate',
+  properties: [
+    {
+      class: 'FObjectProperty',
+      of: 'foam.mlang.predicate.Predicate',
+      name: 'predicate'
+    },
+    {
+      class: 'StringArray',
+      name: 'args'
+    },
+    {
+      class: 'Function',
+      name: 'jsFunc',
+      expression: function(predicate, jsErr) {
+        return function() {
+          if ( ! predicate.f(this) ) return jsErr(this);
+        };
+      }
+    },
+    {
+      class: 'String',
+      name: 'errorString'
+    },
+    {
+      class: 'Function',
+      name: 'jsErr',
+      expression: function(errorString) {
+        return function() { return errorString; };
+      }
+    }
+  ],
+  methods: [
+    function createErrorSlotFor(data) {
+      return this.ExpressionSlot.create({
+        args: this.args.map(a => data[a+'$']),
+        code: this.jsFunc.bind(data)
+      });
+    }
+  ]
+});
+
 foam.CLASS({
   package: 'foam.core',
   name: 'PropertyValidationRefinement',
   refines: 'foam.core.Property',
 
   properties: [
+    {
+      class: 'FObjectArray',
+      of: 'foam.core.ValidationPredicate',
+      name: 'validationPredicates'
+    },
     {
       name: 'validateObj',
-      expression: function(name, label, required) {
+      expression: function(name, label, required, validationPredicates) {
+        if ( validationPredicates.length ) {
+          var args = Object.keys(validationPredicates
+            .map(vp => vp.args)
+            .flat()
+            .reduce((map, a) => {
```

Comment:
You can use foam.Array.unique() insteat of the trick with Object.keys() and .reduce().  Is a bit more readable.

Reply:
Sweet! Done.